File: external/boost_1.55.0/libs/python/test/implicit.py
# Copyright David Abrahams 2004. Distributed under the Boost
# Software License, Version 1.0. (See accompanying
# file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
'''
>>> from implicit_ext import *
>>> x_value(X(42))
42
>>> x_value(42)
42
>>> x = make_x(X(42))
>>> x.value()
42
>>> try: make_x('fool')
... except TypeError: pass
... else: print 'no error'

>>> print x_value.__doc__.splitlines()[1]
x_value( (X)arg1) -> int :

>>> print make_x.__doc__.splitlines()[1]
make_x( (object)arg1) -> X :

>>> print X.value.__doc__.splitlines()[1]
value( (X)arg1) -> int :

>>> print X.set.__doc__.splitlines()[1]
set( (X)arg1, (object)arg2) -> None :

'''

def run(args = None):
    import sys
    import doctest

    if args is not None:
        sys.argv = args
    return doctest.testmod(sys.modules.get(__name__))
    
if __name__ == '__main__':
    print "running..."
    import sys
    status = run()[0]
    if (status == 0): print "Done."
    sys.exit(status)

File: external/boost_1.55.0/libs/python/test/implicit.cpp
// Copyright David Abrahams 2002.
// Distributed under the Boost Software License, Version 1.0. (See
// accompanying file LICENSE_1_0.txt or copy at
// http://www.boost.org/LICENSE_1_0.txt)
#include <boost/python/class.hpp>
#include <boost/python/implicit.hpp>
#include <boost/python/module.hpp>
#include <boost/python/def.hpp>
#include "test_class.hpp"

using namespace boost::python;

typedef test_class<> X;

int x_value(X const& x)
{
    return x.value();
}

X make_x(int n) { return X(n); }


// foo/bar -- a regression for a vc7 bug workaround
struct bar {};
struct foo
{
    virtual ~foo() {}; // silence compiler warnings
    virtual void f() = 0;
    operator bar() const { return bar(); }
};

BOOST_PYTHON_MODULE(implicit_ext)
{
    implicitly_convertible<foo,bar>();
    implicitly_convertible<int,X>();
    
    def("x_value", x_value);
    def("make_x", make_x);

    class_<X>("X", init<int>())
        .def("value", &X::value)
        .def("set", &X::set)
        ;
    
    implicitly_convertible<X,int>();
}

#include "module_tail.cpp"
